File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n = len(s)
        i = maxLen = 0
        freq = {}
        for j in range (n):
            if s[j] in freq:
                freq[s[j]] += 1
            else:
                freq[s[j]] = 1
            windowSize = j - i + 1
            if len(freq) < windowSize:
                freq[s[i]] -= 1
                if freq[s[i]] == 0:
                    del freq[s[i]]
                i += 1
            else:
                maxLen = max(maxLen, windowSize)
        return maxLen

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length(), i = 0, maxLen = 0;
        HashMap<Character, Integer> freq = new HashMap<>();
        for(int j = 0; j < n; ++j) {
            if(freq.containsKey(s.charAt(j))) 
                freq.put(s.charAt(j), freq.get(s.charAt(j)) + 1);
            else 
                freq.put(s.charAt(j), 1);
            int windowSize = j - i + 1;
            if(freq.size() < windowSize) {
                freq.put(s.charAt(i), freq.get(s.charAt(i)) - 1);
                if(freq.get(s.charAt(i)) == 0) 
                    freq.remove(s.charAt(i));
                i++;
            } else {
                maxLen = Math.max(maxLen, windowSize);
            }
        }
        return maxLen;
    }
}